Update Loris to allow converting embedded colour profiles to sRGB on non-JP2 image formats

alexwlchan opened the pull request:
This patch is for addressing https://github.com/wellcometrust/platform/issues/734. We noticed that Loris was stripping the Kodak ProPhoto colour profile from transformed JPEGs, so they looked wrong.

Loris already has support for converting the embedded colour profiles on JP2 images to sRGB. The headline feature of this patch is extending that support to other formats.

This involved making the profile conversion more generic, and I made a bunch of other small changes along the way – mostly aimed at making Loris give more useful errors if you’ve configured this feature incorrectly, so you fall over at startup rather than with an AttributeError during image conversion.

Still to do:

- [x] Add some tests of the embedding process
- [x] (Maybe) Pull the imports into the mixin, so they only run when really needed. Reduces the risk of irrelevant ImportErrors, and would slightly improve performance when this feature wasn’t enabled

Those aside, I think this PR is mostly ready for review.

alexwlchan commented on loris/transforms.py:
```diff
@@ -131,42 +184,34 @@ def _derive_with_pil(self, im, target_fp, image_request, rotate=True, crop=True)
-        if self.map_profile_to_srgb and \
-            ('PIL.ImageCms' not in sys.modules and 'ImageCms' not in sys.modules):
-            logger.warn('Could not import profileToProfile from ImageCms.')
-            logger.warn('Images will not have their embedded color profiles mapped to sSRGB.')
```
I think this warning was unhittable – because `profileToProfile` is imported at the top of this file, if the libraries aren’t installed, we’d never reach this point in the file.

bcail commented:
Thanks, @alexwlchan. Let me know when you're ready for me to review. If you can keep the PR from getting too big with unrelated changes, that'd be great.

alexwlchan committed:
Remove the PIL fallback path for profileToProfile

Loris requires Pillow >= 2.4.0, and 2.4.0 includes PIL.ImageCms.
Thus this fallback path is no longer needed.

alexwlchan committed:
Check that sRGB conversion config is correct before proceeding

alexwlchan committed:
Add tests for validating the sRGB conversion config

alexwlchan committed:
Add a test that we can convert embedded JPEG color profiles

alexwlchan committed:
Move the color profile settings on to _AbstractTransformer

alexwlchan committed:
Add a method for doing the color conversion

alexwlchan committed:
Flip the JP2 transformers to use the new method

alexwlchan committed:
Add a library/symlink required for the OpenJPEG tests

Otherwise the openjpeg commands in the tests hang with the error:

    /repo/bin/Linux/x86_64/opj_decompress: error while loading shared
    libraries: libopenjp2.so.7: cannot open shared object file:
    No such file or directory

alexwlchan committed:
add todo

alexwlchan committed:
Get color conversion working on all transformers

alexwlchan committed:
Add tests for color profile conversion

alexwlchan committed:
Document the sRGB color conversion options

alexwlchan commented:
Okay @bcail, I think this is ready for review. I’ve tidied the commit history and spun out what I can into separate PRs. Thanks for the speedy review on those!

Of particular note:

- I added tests in the style of #351 for JPEG-via-Pillow and JP2-via-OpenJPEG. The latter took a bit of extra work to get right, but I wanted to be sure I hadn’t broken anything.
- I switched from `StringIO` to `BytesIO` because my tests were getting a TypeError otherwise. I’m not entirely sure why that’s the case.
- The original PR was quite mixin and subclass heavy; I’ve tried to pull back from that because I don’t think we need it here.
